ens_mean_recenter/module_mpi_arrange.f90

```diff
@@ -142,174 +142,139 @@ subroutine arrange(this,ncfs_dyn,ncfs_tracer,ncfs_sfc)
     allocate(lvlend(ntotalcore))
     allocate(nx(ntotalcore))
     allocate(ny(ntotalcore))
+    varname=""
+    fileid=0
+    vartype=0
+    lvlbegin=0
+    lvlend=0
+    nx=0
+    ny=0
 !
 ! find number of 3d and 2 d fields
 !
     nlvl2d=0
     nlvl3d=0
+    nlvl3d_small=0
     nlvlmax=0
-    if(ncfs_tracer%num_totalvl > 0) then
-        do k=1,ncfs_tracer%numvar
-           if(ncfs_tracer%num_dim(k) == 3) nlvl3d=nlvl3d+1
-           if(ncfs_tracer%num_dim(k) == 2) nlvl2d=nlvl2d+1
-           if(nlvlmax < ncfs_tracer%dim_3(k)) nlvlmax=ncfs_tracer%dim_3(k)
-        enddo
-    endif
-    if(ncfs_dyn%num_totalvl > 0) then
-        do k=1,ncfs_dyn%numvar
-           if(ncfs_dyn%num_dim(k) == 3) nlvl3d=nlvl3d+1
-           if(ncfs_dyn%num_dim(k) == 2) nlvl2d=nlvl2d+1
-           if(nlvlmax < ncfs_dyn%dim_3(k)) nlvlmax=ncfs_dyn%dim_3(k)
-        enddo
-    endif
-    if(ncfs_sfc%num_totalvl > 0) then
-        do k=1,ncfs_sfc%numvar
-           if(ncfs_sfc%num_dim(k) == 3) nlvl3d=nlvl3d+1
-           if(ncfs_sfc%num_dim(k) == 2) nlvl2d=nlvl2d+1
-           if(nlvlmax < ncfs_sfc%dim_3(k)) nlvlmax=ncfs_sfc%dim_3(k)
+    if(ncfs_all%num_totalvl > 0) then
+        do k=1,ncfs_all%numvar
+           if(ncfs_all%num_dim(k) == 3) then
+              if(ncfs_all%dim_3(k) <= 10) then
+                ! treat small 3d as 2D field
+                nlvl3d_small=nlvl3d_small+ncfs_all%dim_3(k)
+              else
+                nlvl3d=nlvl3d+1
+              endif
+           endif
+           if(ncfs_all%num_dim(k) == 2) nlvl2d=nlvl2d+1
+           if(nlvlmax < ncfs_all%dim_3(k)) nlvlmax=ncfs_all%dim_3(k)
         enddo
     endif
-    write(6,*) 'total 2d level=',nlvl2d,' 3d level=',nlvl3d
+    write(6,*) 'total 2d_level=',nlvl2d,' 3d_level=',nlvl3d, &
+               ' 3d_level_small =',nlvl3d_small
     write(6,*) 'max level =',nlvlmax
 !
 !  check if we have enought cores
 !
-    if( ntotalcore-nlvl2d < nlvl3d ) then
-        write(6,*) 'not enough cores for the paralleli IO',ntotalcore-nlvl2d,nlvl3d
+    if( ntotalcore-nlvl2d-nlvl3d_small < nlvl3d ) then
```

Comment:
Why do we need such a criteria?

Reply:
There are two 3D arrays, 65 level atmosphere and 9 level soil levels. I treat 9 level soil as single level fields (nlvl3d_small).